### ISSUE TITLE
allow inplace operations of expanded array

### ISSUE THREAD

**Issue by florian98765** [pytorch/pytorch#95989]
### 🚀 The feature, motivation and pitch

I am using pytorch for finite difference micromagnetic simulations.
In the simplest case material parameters are float constants, but in general they are represented by a torch.tensor, which stores one value for each cell of the finite difference mesh. 

In order to avoid code branches I expand a constant material value into a constant tensor. This allows to use the same code and have exactly the same type/shape/... without any additional memory.

Unfortunately, using the method it is not possible to modify parts of the expanded vector (e.g. material[0] = 1. not only changes the first elements material, but influences all cells, because the expanded tensor is only a view on a single value).

I would need any method, which catches the inplace operation and replaces the view with a cloned tensor without changing the original reference.

To make it clearer the following code
```
x = torch.Tensor([0.])
y = x.expand(4)
y[0] = 1.
```
currently gives `y = tensor([1., 1., 1., 1.])`, but i would like to get `y = tensor([1., 0., 0., 0.])`

Is this possible somehow?
thanks and best wishes 
Florian

**Comment by andreasfloros** [comment]
You may want to use repeat instead: https://pytorch.org/docs/stable/generated/torch.Tensor.repeat.html

```
x = torch.Tensor([0.])
y = x.repeat(4)
y[0] = 1.
```

gives `y = tensor([1., 0., 0., 0.])`.

**Comment by florian98765** [comment]
No, thats would directly allocate memory for each cell, which i would like to avoid in case of constant material parameters. Only if the constant parameter is changed locally in one cell the full array should be used. Otherwise I would like to use only a single double value and a corresponding view, in order to minimize memory consumption. 

So before the inplace setting is done, the tensor should be cloned, but in a way that the reference of the original tensor is not changed.

**Comment by andreasfloros** [comment]
I see. So ideally you would like to store the cell indices and the corresponding values to avoid storing the full tensor?
Maybe you should be working with a kind of sparse tensor: https://pytorch.org/docs/stable/sparse.html#sparse-compressed-tensors

**Comment by florian98765** [comment]
No. I am sorry for the unclear description. I will try it again:
For example we could have the following simple application cases for a simple 1D model with material parameter A, and `N=4` cells:
1.) constant A=1 -> `A = torch.Tensor([1.]).expand(N)`
2.) variable material: `A = torch.Tensor([1.,2.,3.,4.])`

in the first case only a singe float should be stored. 
in the second case N floats need to be stored. In both cases A has the same shape and could be used similarly. 
Since the material could be changed during the simulation, there is also an additional case:
3.) we start with a constant material `A = torch.Tensor([1.]).expand(N)`. And then we want to modify only `A[1] = 2`. Unfortunately, this does not lead to the expected result, but sets all entries to 2., since the underlying memory of each element points to a single float. 

Of course, one could do 
```
A = A.clone() # this would create a full array
A[1] = 2
```
but in my case, e.g. the material is stored inside of a dictionary. If one gets A via `material["A"]`, nobody can hinder the user from resetting one element of A. But this would then lead to the unexpected behaviour. 

I hope the problem is clearer now. 
I am not sure if it is even possible to change the underlying data of a tensor. I would also be happy with a workaround, as long as it is transparent for the user.

thanks for any ideas.
best wishes 
Florian

**Comment by florian98765** [comment]
ok, i understand the problem. maybe it would be possible to add an explicit method that can be used to perform and `inplace_clone` of an expanded tensor? I could check in may code, if the tensor is expanded and call the `inplicit_clone` in this case.

**Comment by florian98765** [comment]
yes, sorry. Let's make an example:
```
material["A"] = torch.Tensor([1.]).expand(N) # start with a constant material

# inplace updated does not work as expected 
material["A"][0] = 2. # this gives a constant 2. in all elements

# here comes the inplace_clone which transforms the expanded tensor into a normal tensor 
material["A"].inplace_clone()
# now one element of the tensor could be changed 
material["A"][0] = 2.  # should give [2., 1., 1., ...]
```
The point is that the user can get `material["A"]` and nothing hinders him to change a single element, but with unexpected results if the tensor was expanded. It the `inplace_clone` function was available one could maybe use overloading of the `torch.Tensor` to clone the tensor if needed.

**Comment by ngimel** [comment]
`contiguous()` has the same functionality?

**Comment by florian98765** [comment]
hhmm, but i thought this is not in inplace operation. It just returns a copy of the tensor like `clone`.
So I would have to update the material dictionary in the upper example since the reference to the tensor changes. 
Something like 
```
A = material["A"]
# now i want to change one element of A
A = A.clone()
A[0] = 2.
# but now i need to reset the material in the dictionary, otherwise the changes would only apply to the local copy of the material
material["A"] = A
```
The problem with this solutions is that i like to hide the clone command from the user.

I'm not sure if it is possible to internally change the memory of a tensor.

**Comment by florian98765** [comment]
just found `x.set_`. I hope that this should do the job.

**Comment by ngimel** [comment]
you cannot autograd through `set_`, but if you don't need gradient functionality that's fine.
